File: CSharp/SpecStudioExamplesTestsCSharp/common/FilterValueString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class FilterValueString
    {
        public string value;

        public FilterValueString(string value)
        {
            this.value = value;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static FilterValueString FromText(string text)
        {
            var parts = Tokens.Require(text, 1, "FilterValue");
            return new FilterValueString(parts[0]);
        }

        public FilterValueTyped ToFilterValueTyped()
        {
            return new FilterValueTyped(
                new IDForm(this.value)
            );
        }

        public override string ToString()
        {
            return Tokens.Token(value);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not FilterValueString other) return false;
            return (DNCString == this.value || DNCString == other.value || object.Equals(this.value, other.value));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.value);
            return h.ToHashCode();
        }
    }
}

File: Java/SpectableJavaExamplesTest/src/test/java/spectable/common/FilterValueString.java
package spectable.common;

import java.util.Objects;
import production.*;
import records.*;
import calculator.*;

public class FilterValueString {
    private static final String DNCString = "?DNC?";

    public String value;

    public FilterValueString(String value) {
        this.value = value;
    }

    /** Builds from the text form, e.g. Money as "25 USD". */
    public static FilterValueString fromText(String text) {
        java.util.List<String> parts = Tokens.require(text, 1, "FilterValue");
        return new FilterValueString(parts.get(0));
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof FilterValueString)) return false;
        FilterValueString that = (FilterValueString) o;
        return (DNCString.equals(value) || DNCString.equals(that.value) || Objects.equals(value, that.value));
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }

    @Override
    public String toString() {
        return Tokens.token(value);
    }
}
